constraint-solver/src/quadratic_symbolic_expression.rs

```diff
@@ -208,7 +208,80 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq> QuadraticSymbolicExpression<T,
         }
     }
 
-    // TODO implement substitute_by_unknown where substitution is a QSE.
+    /// Substitute a variable by an unknown expression (a QuadraticSymbolicExpression).
+    pub fn substitute_by_unknown(
+        &mut self,
+        variable: &V,
+        substitution: &QuadraticSymbolicExpression<T, V>,
+    ) {
+        if !self.referenced_variables().any(|v| v == variable) {
+            return;
+        }
+
+        match &self.constant {
+            SymbolicExpression::Symbol(sym, _) if sym == variable => {
+                self.constant = SymbolicExpression::from(T::zero());
+                *self += substitution.clone();
+            }
+            _ => {
+                if let Some(mut qse) =
+                    symbolic_expression_to_quadratic_symbolic_expression(&self.constant)
+                {
+                    qse.substitute_by_unknown(variable, substitution);
+                    self.constant = SymbolicExpression::from(T::zero());
+                    *self += qse;
+                }
+            }
+        }
```

Comment:
If the variable appears here, it means that it is known, so doing this would turn it back to unknown. I think that's not wha we want or need, is it?

Reply:
Fixed.